job/src/main/java/org/jenkinsci/plugins/workflow/job/WorkflowJobProperty.java

```diff
@@ -0,0 +1,43 @@
+/*
+ * The MIT License
+ *
+ * Copyright 2015 CloudBees, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package org.jenkinsci.plugins.workflow.job;
+
+import hudson.model.JobProperty;
+import hudson.security.ACL;
+import javax.annotation.Nonnull;
+
+/**
+ * Job property which can specially affect behavior of the project.
+ */
+public abstract class WorkflowJobProperty extends JobProperty<WorkflowJob> {
+
+    /**
+     * Allows a property to control permissions related to the job.
+     */
+    public @Nonnull ACL decorateACL(@Nonnull ACL acl) {
```

Comment:
Perhaps. I am not completely comfortable with doing this via a `JobProperty`-based API, just following a convention based on behavioral hooks in `JobProperty` itself, not to mention `FolderProperty`. The more flexible approach is to use an `@Extension` which can choose to look for a particular `JobProperty` or use any other criterion. I am open to switching to that idiom here. (Though it needs to be verified that there is minimal overhead in iterating the extensions, so as not to adversely impact jobs not using the property. I _think_ the extension lists get cached effectively.)

For the future, if we are talking about things to be baked into core, I think a more general set of APIs is needed to allow external customization of ACL decisions. CJP-2326 tracks some thoughts on this, which if accepted would need to be written up into a more concrete proposal in JENKINS.